### PR TITLE
When moving a shard to a new node ensure there is enough space

### PR DESCRIPTION
**Pull request by SaitTalhaNisanci**
DESCRIPTION: Checks free disk space before moving a shard

TODO:
- [x] Add some more tests
- [x] Add isolation tests
- [x] Decide on GUC names
- [x] Decide on the default values for GUCs

### CODE REVIEW

File: src/backend/distributed/operations/repair_shards.c [hunk 1]
```diff
@@ -287,7 +413,119 @@ citus_move_shard_placement(PG_FUNCTION_ARGS)
...
+			ereport(ERROR, (
+						errmsg(
+							"shards were not removed after trying %d times, giving up",
+							checkSpaceTries),
+						errhint("Try increasing citus.move_shard_defer_drop_tries")
+						));
+		}
+
+		/* wait the timeout and try again*/
+		pg_usleep(WaitForDeferShardRetryTimeInSec * 1000 * 1000);
```
Comment (onderkalaci) [inline src/backend/distributed/operations/repair_shards.c:488]: sleeping might not be the most optimal way because with it you cannot read the signals that happen before the loop finishes. See `WaitForMiliseconds ()` in the enterprise repo

File: src/backend/distributed/operations/repair_shards.c [hunk 2]
```diff
@@ -158,6 +176,111 @@ master_copy_shard_placement(PG_FUNCTION_ARGS)
...
+	if (ForceDiskAvailable != -1)
+	{
+		diskAvailable = ForceDiskAvailable;
+	}
+	else
+	{
+		diskAvailable = ExecuteRemoteInt64Command(connection,
+												  "select pg_catalog.citus_disk_available()");
+	}
+	if (ForceDiskSize != -1)
```
Comment (onderkalaci) [inline src/backend/distributed/operations/repair_shards.c:249]: as you typically suggest, can we use macros instead of constant integers for the comparisons?

File: src/backend/distributed/operations/repair_shards.c (hunk 1 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/operations/repair_shards.c:488]: WaitForMilliseconds makes sense, I see some other usages of pg_usleep in the codebase. Is there a downside of using `WaitForMilliseconds` compared to `pg_usleep`

Comment (onderkalaci) [reply src/backend/distributed/operations/repair_shards.c:488]: As far as I know, no there is not a downside of using it. Still, would be good to double check the parameters that we pass to `WaitLatch` in `WaitForMilliseconds`

File: src/backend/distributed/operations/repair_shards.c [hunk 3]
```diff
@@ -83,15 +86,30 @@ static void UpdateColocatedShardPlacementMetadataOnWorkers(int64 shardId,
...
 PG_FUNCTION_INFO_V1(master_move_shard_placement);
+PG_FUNCTION_INFO_V1(citus_disk_available);
+PG_FUNCTION_INFO_V1(citus_disk_size);
 
 
 bool DeferShardDeleteOnMove = false;
+int ForceDiskAvailable = -1;
+int ForceDiskSize = -1;
+
+int WaitForDeferShardRetryTimeInSec = 30;
```
Comment (SaitTalhaNisanci) [inline src/backend/distributed/operations/repair_shards.c:109]: Are these defaults good?

File: src/backend/distributed/operations/repair_shards.c (hunk 2 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/operations/repair_shards.c:249]: Makes sense, I think ForceDiskSizeInBytes is also better than this so that we know the metric when we look at the variable name

File: src/backend/distributed/operations/repair_shards.c [hunk 4]
```diff
@@ -158,6 +176,111 @@ master_copy_shard_placement(PG_FUNCTION_ARGS)
...
+	}
+	else
+	{
+		diskSize = ExecuteRemoteInt64Command(connection,
+											 "select pg_catalog.citus_disk_size()");
+	}
+
+	if (diskAvailable < colocationSize)
+	{
+		appendStringInfo(errorDetail, "actual space %ld, required space %ld",
```
Comment (SaitTalhaNisanci) [inline src/backend/distributed/operations/repair_shards.c:261]: I am adding the units here "bytes" in the error message so that it is clear how much space there is.

Comment (SaitTalhaNisanci) [reply src/backend/distributed/operations/repair_shards.c:261]: Also in case the actual space is "forced", it seems better to say this in the error message for easier identification of problems

File: src/backend/distributed/sql/citus--9.1-1--9.2-1.sql [hunk 5]
```diff
@@ -83,3 +83,17 @@ CREATE TRIGGER pg_dist_rebalance_strategy_enterprise_check_trigger
 
 #include "udfs/citus_prepare_pg_upgrade/9.2-1.sql"
 #include "udfs/citus_finish_pg_upgrade/9.2-1.sql"
+
+CREATE OR REPLACE FUNCTION pg_catalog.citus_disk_available()
```
Comment (SaitTalhaNisanci) [inline src/backend/distributed/sql/citus--9.1-1--9.2-1.sql:87]: these should be moved to their own folder

File: src/backend/distributed/operations/repair_shards.c [hunk 6]
```diff
@@ -287,7 +429,120 @@ citus_move_shard_placement(PG_FUNCTION_ARGS)
...
+
+/*
+ * citus_disk_available returns the number of bytes of available diskspace that
+ * are on the given tablespace.
+ */
+Datum
+citus_disk_available(PG_FUNCTION_ARGS)
+{
+	struct statvfs diskstats;
+	int error = statvfs("base/", &diskstats);
```
Comment (marcocitus) [inline src/backend/distributed/operations/repair_shards.c:518]: fwiw, this PR also adds some similar functions:

https://github.com/citusdata/citus/pull/4383/files#diff-a0fe2d7cbca8cd88f0580ff39c74be56eb992920a2b2acd1405e71cf4095a0bfR239

File: src/backend/distributed/shared_library_init.c [hunk 7]
```diff
@@ -924,6 +925,63 @@ RegisterCitusConfigVariables(void)
...
+			"some more free space so that there is enough space for the shard move."
+			"This setting determines the duration of sleep (in second) between the deletion operations."),
+		&WaitForDeferShardRetryTimeInSec,
+		30, 0, INT_MAX,
+		PGC_USERSET,
+		GUC_UNIT_S,
+		NULL, NULL, NULL);
+
+	DefineCustomRealVariable(
+		"citus.desired_percent_disk_available_after_move",
```
Comment (marcocitus) [inline src/backend/distributed/shared_library_init.c:972]: we're adding quite a lot of GUCs here. Do we need all of them? Do we need all of them to be user-visible?

Comment (marcocitus) [reply src/backend/distributed/shared_library_init.c:972]: I know we're adding this because we don't really know what to set the values to, but in practice these types of settings are basically never used. Can we simplify it to a) you use disk-space protection and we decide good values, b) you don't use disk-space protection ?

Comment (SaitTalhaNisanci) [reply src/backend/distributed/shared_library_init.c:972]: I think it makes sense to simplify this, with what you suggested, we would have 1 user-visible GUC(we will also need force_disk_available and force_disk_size but they are not user-visible).

File: src/backend/distributed/operations/repair_shards.c [hunk 8]
```diff
@@ -287,7 +429,120 @@ citus_move_shard_placement(PG_FUNCTION_ARGS)
...
+														sourceNodeName,
+														sourceNodePort);
+
+	uint32 connectionFlag = 0;
+	MultiConnection *connection = GetNodeConnection(connectionFlag, targetNodeName,
+													targetNodePort);
+	int checkSpaceTries = 0;
+	while (true)
+	{
+		bool droppedAllOldShards = false;
```
Comment (onderkalaci) [inline src/backend/distributed/operations/repair_shards.c:451]: What is the value that retrying to drop adds here? Can't we just try once and throw error if it fails?

That'd simplify the logic quite a bit and avoid 1 GUC: `WaitForDeferShardsMaxTries`?

Comment (onderkalaci) [reply src/backend/distributed/operations/repair_shards.c:451]: even 2 gucs I think: `WaitForDeferShardRetryTimeInSec`

File: src/backend/distributed/connection/remote_commands.c [hunk 9]
```diff
@@ -372,6 +372,20 @@ ExecuteCriticalRemoteCommandList(MultiConnection *connection, List *commandList)
...
+
+
+/*
+ * ExecuteCriticalRemoteCommandWithResult executes a remote command that is
+ * critical to the transaction. If the command fails then the transaction
+ * aborts. This function also allows you to get the result of the query back.
+ * If NULL is passed as the result, the result will be automatically cleared.
+ */
+void
+ExecuteCriticalRemoteCommandWithResult(MultiConnection *connection, const char *command,
```
Comment (onderkalaci) [inline src/backend/distributed/connection/remote_commands.c:387]: This API seems a little unexpected. `GetRemoteCommandResult` may need be called multiple times to fetch the results. Here you can safely call for your purpose because you expect a single tuple result. 

So, I'd rather have a specific function for that instead of having a generic `ExecuteCriticalRemoteCommandWithResult` such as `TotalRelationSizeForSubscription`

File: src/backend/distributed/shared_library_init.c [hunk 10]
```diff
@@ -924,6 +925,63 @@ RegisterCitusConfigVariables(void)
...
 
+	DefineCustomBoolVariable(
+		"citus.check_available_space_before_move",
+		gettext_noop("When enabled will check free disk space before a shard move"),
+		gettext_noop(
+			"Free disk space will be checked when this setting is enabled before each shard move."
+			"See citus.delete_old_shards_sleep_time and citus.delete_old_shards_max_tries"),
+		&CheckAvailableSpaceBeforeMove,
+		true,
+		PGC_USERSET,
```
Comment (onderkalaci) [inline src/backend/distributed/shared_library_init.c:936]: I think we can mark this as NO_SHOW Guc.

File: src/backend/distributed/operations/repair_shards.c (hunk 6 above)
Comment (onderkalaci) [reply src/backend/distributed/operations/repair_shards.c:518]: I think the API Marco noted looks better, a single UDF call gets both of the disk stats at once.

Also, I have not dived on the internals of the functions, but @SaitTalhaNisanci can you make sure that if they are the same? If not, can you asses which one is more accurate?

File: src/backend/distributed/operations/repair_shards.c [hunk 11]
```diff
@@ -158,6 +178,125 @@ master_copy_shard_placement(PG_FUNCTION_ARGS)
 }
 
 
+/*
+ * ColocationSizeInBytes returns the size in bytes of a set of colocated tables.
```
Comment (onderkalaci) [inline src/backend/distributed/operations/repair_shards.c:182]: Why the name starts with `colocation`? We give a list of shards and it calculates the shards?

File: src/backend/distributed/shared_library_init.c [hunk 12]
```diff
@@ -939,6 +997,33 @@ RegisterCitusConfigVariables(void)
...
+					 "size available on the worker. If it's -1 the actual size"
+					 "of the worker will be used. "),
+		&ForceDiskAvailableInBytes,
+		-1, -1, INT_MAX,
+		PGC_USERSET,
+		GUC_NO_SHOW_ALL,
+		NULL, NULL, NULL);
+
+	DefineCustomIntVariable(
+		"citus.force_disk_size",
```
Comment (onderkalaci) [inline src/backend/distributed/shared_library_init.c:1014]: What is the use case for `force_disk_size`? In other words, why would anyone need this ever?

File: src/backend/distributed/utils/time_utils.c [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+	/* wait until timeout, or until somebody wakes us up */
+	int rc = WaitLatch(MyLatch, latchFlags, timeout, PG_WAIT_EXTENSION);
+
+	/* emergency bailout if postmaster has died */
+	if (rc & WL_POSTMASTER_DEATH)
+	{
+		proc_exit(1);
+	}
+
+	if (rc & WL_LATCH_SET)
```
Comment (onderkalaci) [inline src/backend/distributed/utils/time_utils.c:29]: Shouldn't we need to read the config file here, at least for PG 13?

File: src/backend/distributed/shared_library_init.c [hunk 14]
```diff
@@ -924,6 +925,63 @@ RegisterCitusConfigVariables(void)
...
+	DefineCustomRealVariable(
+		"citus.desired_percent_disk_available_after_move",
+		gettext_noop(
+			"Sets how many percentage of free disk space should be after a shard move"),
+		gettext_noop(
+			"This setting controls how much free space should be available after a shard move."
+			"If the free disk space will be lower than this parameter, then shard move will result in"
+			"an error."),
+		&DesiredPercentFreeAfterMove,
+		5.0, 0.0, 100.0,
```
Comment (onderkalaci) [inline src/backend/distributed/shared_library_init.c:980]: As far as I remember, HSC converts the system to read-only if disk space left is less than 5%. So, in that regard, the smallest allowed should probably be 6%. But, lets keep the smallest allowed as 0, just for emergency cases.

But, I think the default can be 10% (or even more?).

File: src/backend/distributed/operations/shard_cleaner.c [hunk 15]
```diff
@@ -95,17 +100,23 @@ TryDropMarkedShards(bool waitForCleanupLock)
...
- * wait for this lock or returns with a warning.
+ * wait for this lock or error out.
+ *
+ * The function returns true if all shards were successfuly dropped and fills
+ * removedShardCount with the number of shards that were dropped.
  */
-static int
-DropMarkedShards(bool waitForCleanupLock)
+bool
+DropMarkedShards(bool waitForCleanupLock, int *removedShardCount)
```
Comment (onderkalaci) [inline src/backend/distributed/operations/shard_cleaner.c:109]: do we need this changes if we remove the retry loop?

File: src/backend/distributed/shared_library_init.c [hunk 16]
```diff
@@ -939,6 +997,33 @@ RegisterCitusConfigVariables(void)
 		GUC_UNIT_MS,
 		NULL, NULL, NULL);
 
+	DefineCustomIntVariable(
```
Comment (onderkalaci) [inline src/backend/distributed/shared_library_init.c:1000]: I really cannot use case for this GUC as well. Can you explain why do you think needed?

File: src/backend/distributed/operations/repair_shards.c [hunk 17]
```diff
@@ -158,6 +178,125 @@ master_copy_shard_placement(PG_FUNCTION_ARGS)
...
+
+/*
+ * CheckSpaceConstraints checks there is enough space to place the colocation
+ * on the node that the connection is connected to.
+ */
+static
+DeferredErrorMessage *
+CheckSpaceConstraints(MultiConnection *connection, int64 colocationSizeInBytes)
+{
+	if (!CheckAvailableSpaceBeforeMove)
```
Comment (onderkalaci) [inline src/backend/distributed/operations/repair_shards.c:236]: I expected to see this check at the start of `EnsureEnoughDiskSpaceForShardMove`. Is there a reason for us defer it until here?

File: src/backend/distributed/shared_library_init.c (hunk 12 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/shared_library_init.c:1014]: this is used in testing, in reality noone would need this I think. With this we set the "disk size". This is not visible to users

File: src/backend/distributed/operations/repair_shards.c (hunk 6 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/operations/repair_shards.c:518]: They seem quite similar, I will go with the single UDF version. Also what these fields mean:

image

f_bsize and f_frsize seems to be the same almost all the time so I guess there isn't much difference in using either

File: src/backend/distributed/operations/repair_shards.c (hunk 8 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/operations/repair_shards.c:451]: As we talked in private I think this doesn't add much value here. So simplifying makes sense, I am removing the 2 GUCs

File: src/backend/distributed/operations/repair_shards.c (hunk 11 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/operations/repair_shards.c]: I am changing it as `ShardListSizeInBytes`

File: src/backend/distributed/utils/time_utils.c (hunk 13 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/utils/time_utils.c]: Yes, but we don't need this file anymore since we are removing the retry loop. Also this is fixed on enterprise in case you want to review :) https://github.com/citusdata/citus-enterprise/pull/583

File: src/backend/distributed/operations/shard_cleaner.c (hunk 15 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/operations/shard_cleaner.c]: I think the reason for this change was that when `DropMarkedShards` error we don't want to return just "-1" to the callers but the actual number of shards that were dropped. But with the current version I don't think `DropMarkedShards` can error this number incorrectly because we have `TryDropShard` and it catches errors already. Therefore we can go back to the previous version

File: src/backend/distributed/shared_library_init.c (hunk 16 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/shared_library_init.c]: This is used for testing

File: src/backend/distributed/shared_library_init.c (hunk 10 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/shared_library_init.c:949]: Don't we want users to see this setting? I think if we simplify the gucs to https://github.com/citusdata/citus/pull/4929#discussion_r623971584 then it might make sense to make this user visible, what do you think? 

Maybe we can make `DesiredPercentFreeAfterMove` invisible to users and decide the value ourselves as Marco suggested. 

So the only user visible setting could be `citus.check_available_space_before_move`, from user point of the view, the interaction might be easier/simpler this way. We can keep `citus.desired_percent_disk_available_after_move` as a GUC so that if we are not okay with the default value we can change it in production.

File: src/backend/distributed/connection/remote_commands.c (hunk 9 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/connection/remote_commands.c]: These changes are gone after removing the "retry loop"

File: src/backend/distributed/metadata/metadata_utility.c [hunk 18]
```diff
@@ -98,12 +100,138 @@ static void AppendShardSizeMinMaxQuery(StringInfo selectQuery, uint64 shardId,
...
+ */
+Datum
+citus_local_disk_space_stats(PG_FUNCTION_ARGS)
+{
+	uint64 availableBytes = 0;
+	uint64 totalBytes = 0;
+
+	if (GetLocalDiskSpaceStats(&availableBytes, &totalBytes))
+	{
+		ereport(WARNING, (errmsg("could not get disk space")));
```
Comment (onderkalaci) [inline src/backend/distributed/metadata/metadata_utility.c:150]: On my mac, I get the following output. Do you have any clue on this?
```

[local] onderkalaci@postgres:5432-24984=# select * from  citus_local_disk_space_stats();
WARNING:  could not get disk space
 available_disk_size | total_disk_size 
---------------------+-----------------
        171613466624 |    499963174912
(1 row)

Time: 0.530 ms
[local] onderkalaci@postgres
```

Comment (onderkalaci) [reply src/backend/distributed/metadata/metadata_utility.c:150]: I guess we should have if (!GetLocalDiskSpaceStats()) warning

File: src/backend/distributed/metadata/metadata_utility.c [hunk 19]
```diff
@@ -98,12 +100,138 @@ static void AppendShardSizeMinMaxQuery(StringInfo selectQuery, uint64 shardId,
...
+
+	int queryResult = ExecuteOptionalRemoteCommand(connection, sizeQuery, &result);
+	if (queryResult != 0 || !IsResponseOK(result) || PQntuples(result) != 1)
+	{
+		ereport(WARNING, (errcode(ERRCODE_CONNECTION_FAILURE),
+						  errmsg("cannot get the disk space statistics for node %s:%d",
+								 connection->hostname, connection->port)));
+
+		PQclear(result);
+		ClearResults(connection, raiseErrors);
```
Comment (onderkalaci) [inline src/backend/distributed/metadata/metadata_utility.c:217]: minor: declare raiseErrors here

File: src/backend/distributed/metadata/metadata_utility.c (hunk 18 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/metadata/metadata_utility.c:150]: Do you get the same warning in https://github.com/citusdata/citus/pull/4383?

Comment (SaitTalhaNisanci) [reply src/backend/distributed/metadata/metadata_utility.c:150]: I think you will get the same warning there because this check should be:

```
if (!GetLocalDiskSpaceStats(&availableBytes, &totalBytes))
```

Not

```
if (GetLocalDiskSpaceStats(&availableBytes, &totalBytes))
```

File: src/backend/distributed/operations/repair_shards.c [hunk 20]
```diff
@@ -158,6 +173,120 @@ master_copy_shard_placement(PG_FUNCTION_ARGS)
...
+							 diskAvailableInBytes, colocationSizeInBytes);
+		}
+		else
+		{
+			appendStringInfo(errorDetail,
+							 "actual space %ld bytes, required space %ld bytes",
+							 diskAvailableInBytes, colocationSizeInBytes);
+		}
+
+		return DeferredError(ERRCODE_INSUFFICIENT_RESOURCES,
```
Comment (onderkalaci) [inline src/backend/distributed/operations/repair_shards.c:268]: Several users do not see `details`. Instead, can we move the detail part to the message?

File: src/backend/distributed/operations/repair_shards.c [hunk 21]
```diff
@@ -287,7 +419,57 @@ citus_move_shard_placement(PG_FUNCTION_ARGS)
...
+	RaiseDeferredError(temporaryError, WARNING);
+
+	/*
+	 * Only run shard cleanup when the first round of checks has failed
+	 * otherwise simply let the automatic cleanup do its job.
+	 *
+	 * NOTE: this tries to drop all marked shards, not only on the shard
+	 * that we're trying to move to.
+	 */
+	ereport(NOTICE, (errmsg("trying to drop old shards to resolve issue...")));
```
Comment (onderkalaci) [inline src/backend/distributed/operations/repair_shards.c:457]: `to resolve issue` sounds vague and can be confusing.  Also, can we return early if there are no `SHARD_STATE_TO_DELETE`?

And, if we have, we Can we tell something like: `found %d old shards on node XXX. Trying to drop old shards to allow the shard move operation to continue.`

File: src/backend/distributed/operations/repair_shards.c [hunk 22]
```diff
@@ -287,7 +419,57 @@ citus_move_shard_placement(PG_FUNCTION_ARGS)
...
+
+	bool waitForCleanupLock = true;
+	DropMarkedShards(waitForCleanupLock);
```
Comment (onderkalaci) [inline src/backend/distributed/operations/repair_shards.c:460]: maybe check the return value of `DropMarkedShards` and if it is zero, print an error and return early. This is to give a more information to the user

File: src/backend/distributed/operations/shard_cleaner.c [hunk 23]
```diff
@@ -75,6 +75,7 @@ TryDropMarkedShards(bool waitForCleanupLock)
 		/* rethrow as WARNING */
 		edata->elevel = WARNING;
 		ThrowErrorData(edata);
+		return -1;
```
Comment (onderkalaci) [inline src/backend/distributed/operations/shard_cleaner.c:78]: why do we return here and return `-1`?

File: src/backend/distributed/operations/shard_cleaner.c [hunk 24]
```diff
@@ -95,17 +96,18 @@ TryDropMarkedShards(bool waitForCleanupLock)
...
+int
 DropMarkedShards(bool waitForCleanupLock)
 {
 	int removedShardCount = 0;
 	ListCell *shardPlacementCell = NULL;
 
 	if (!IsCoordinator())
 	{
-		return removedShardCount;
+		return false;
```
Comment (onderkalaci) [inline src/backend/distributed/operations/shard_cleaner.c:110]: returning false seems unexpected here as the return type is int

File: src/backend/distributed/shared_library_init.c (hunk 12 above)
Comment (onderkalaci) [reply src/backend/distributed/shared_library_init.c:981]: Having a GUC and complexity in the code seems can be avoided by changing the function during the test:
```SQL

create or replace function pg_catalog.citus_local_disk_space_stats(OUT  available_disk_size bigint, OUT total_disk_size bigint)

as $BODY$
begin
    select 10000 into available_disk_size;
select 200000 into total_disk_size;

end
$BODY$ language plpgsql;
```

and then replace back to the original definition for the rest of the test.

Is this enough to avoid 2 GUCs and the related code?

File: src/test/regress/expected/isolation_rebalancer_deferred_drop.out [hunk 25]
```diff
@@ -55,6 +55,28 @@ step s1-commit:
...
+
+step s2-move-placement:
+        SET citus.defer_drop_after_shard_move TO ON;
+        SET citus.force_disk_available = 20;
+     SELECT master_move_shard_placement((SELECT * FROM selected_shard), 'localhost', 57637, 'localhost', 57638);
+ <waiting ...>
+step s1-commit: 
+    COMMIT;
+
+step s2-move-placement: <... completed>
```
Comment (onderkalaci) [inline src/test/regress/expected/isolation_rebalancer_deferred_drop.out:77]: is this test giving the expected output? I guess the previous step should have been from 57638 to 57637 ?

And, why is this test an isolation test? I couldn't follow

File: src/backend/distributed/operations/repair_shards.c [hunk 26]
```diff
@@ -158,6 +173,120 @@ master_copy_shard_placement(PG_FUNCTION_ARGS)
...
+			appendStringInfo(errorDetail,
+							 "actual space %ld bytes, required space %ld bytes",
+							 diskAvailableInBytes, colocationSizeInBytes);
+		}
+
+		return DeferredError(ERRCODE_INSUFFICIENT_RESOURCES,
+							 "not enough space to move shard",
+							 errorDetail->data, NULL);
+	}
+	int64 newDiskAvailableInBytes = diskAvailableInBytes - colocationSizeInBytes;
```
Comment (onderkalaci) [inline src/backend/distributed/operations/repair_shards.c:272]: maybe: `newDiskAvailableInBytes ` -> `diskAvailableInBytesAfterShardMove `

File: src/backend/distributed/operations/repair_shards.c [hunk 27]
```diff
@@ -158,6 +173,120 @@ master_copy_shard_placement(PG_FUNCTION_ARGS)
...
+	int64 desiredNewDiskAvailableInBytes = diskSizeInBytes *
+										   (DesiredPercentFreeAfterMove / 100);
+	if (newDiskAvailableInBytes < desiredNewDiskAvailableInBytes)
+	{
+		appendStringInfo(errorDetail, "actual available space after move %ld bytes, "
+									  "desired available space after move %ld bytes",
+						 newDiskAvailableInBytes, desiredNewDiskAvailableInBytes);
+		return DeferredError(ERRCODE_INSUFFICIENT_RESOURCES,
```
Comment (onderkalaci) [inline src/backend/distributed/operations/repair_shards.c:280]: this error message feels like we moved the shards and we don't have enough space. Can you make it clear that this has not happened, but would happen if we try to move.

And, the hint also can be more descriptive.

File: src/backend/distributed/shared_library_init.c (hunk 14 above)
Comment (onderkalaci) [reply src/backend/distributed/shared_library_init.c:962]: Ping on this, I think we should consider making the default at least 10.

**Review by onderkalaci — COMMENTED** [review]
I think we are going in the right direction, though I have not tested the patch yet

File: src/backend/distributed/operations/repair_shards.c (hunk 21 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/operations/repair_shards.c:457]: Makes sense,

I have one minor suggestion though, we have this at cluster level rather than node level ( of course we can get this at node level as well) but it is simpler to say something like the following:
```
ereport(NOTICE, (errmsg("found %d old shards, trying to drop them to allow the shard move operation to continue", numOfShardsToBeDeleted)));
```

Do you think that is enough apart from that we return early?

File: src/backend/distributed/operations/shard_cleaner.c (hunk 23 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/operations/shard_cleaner.c:78]: With the current structure, we don't need to return here I think because if we error, the return value from this function will be zero and we only check if the return value is positive in maintanence part:

```
			if (numberOfDroppedShards > 0)
			{
				ereport(LOG, (errmsg("maintenance daemon dropped %d distributed "
									 "shards previously marked to be removed",
									 numberOfDroppedShards)));
			}
```

File: src/backend/distributed/operations/repair_shards.c (hunk 27 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/operations/repair_shards.c:280]: Here I am also thinking of adding how much the estimated colocation size is so that users can have an idea of how much space is needed

File: src/backend/distributed/operations/repair_shards.c (hunk 22 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/operations/repair_shards.c]: Also added information added how many shards are dropped

File: src/backend/distributed/operations/repair_shards.c [hunk 28]
```diff
@@ -83,16 +86,26 @@ static void UpdateColocatedShardPlacementMetadataOnWorkers(int64 shardId,
...
+											  char *targetNodeName, uint32
+											  targetNodePort);
 
 /* declarations for dynamic loading */
 PG_FUNCTION_INFO_V1(citus_copy_shard_placement);
 PG_FUNCTION_INFO_V1(master_copy_shard_placement);
 PG_FUNCTION_INFO_V1(citus_move_shard_placement);
 PG_FUNCTION_INFO_V1(master_move_shard_placement);
 
+#define NOT_SET -1
```
Comment (onderkalaci) [inline src/backend/distributed/operations/repair_shards.c:102]: not used anymore?

File: src/backend/distributed/operations/repair_shards.c [hunk 29]
```diff
@@ -287,7 +400,78 @@ citus_move_shard_placement(PG_FUNCTION_ARGS)
...
+		return;
+	}
+	uint64 colocationSizeInBytes = ShardListSizeInBytes(colocatedShardList,
+														sourceNodeName,
+														sourceNodePort);
+
+	uint32 connectionFlag = 0;
+	MultiConnection *connection = GetNodeConnection(connectionFlag, targetNodeName,
+													targetNodePort);
+	DeferredErrorMessage *temporaryError =
```
Comment (onderkalaci) [inline src/backend/distributed/operations/repair_shards.c:423]: minor: `temporaryError` -> `deferredError`

File: src/backend/distributed/operations/repair_shards.c [hunk 30]
```diff
@@ -287,7 +400,78 @@ citus_move_shard_placement(PG_FUNCTION_ARGS)
...
+	 * otherwise simply let the automatic cleanup do its job.
+	 *
+	 * NOTE: this tries to drop all marked shards, not only on the shard
+	 * that we're trying to move to.
+	 */
+	ereport(NOTICE, (errmsg(
+						 "found %d old shard(s), trying to drop them to allow the shard move operation to continue",
+						 numOfShardsToBeDeleted)));
+
+	bool waitForCleanupLock = true;
```
Comment (onderkalaci) [inline src/backend/distributed/operations/repair_shards.c:453]: Do we really want to wait? Not sure, I'm thinking about it

File: src/backend/distributed/operations/shard_rebalancer.c [hunk 31]
```diff
@@ -444,62 +445,27 @@ citus_shard_cost_by_disk_size(PG_FUNCTION_ARGS)
...
-	List *colocatedShardList = ColocatedNonPartitionShardIntervalList(shardInterval);
-	StringInfo tableSizeQuery = GenerateSizeQueryOnMultiplePlacements(colocatedShardList,
-																	  TOTAL_RELATION_SIZE,
-																	  optimizePartitionCalculations);
-
-	MultiConnection *connection = GetNodeConnection(connectionFlag, workerNodeName,
-													workerNodePort);
-	int queryResult = ExecuteOptionalRemoteCommand(connection, tableSizeQuery->data,
-												   &result);
+	List *colocatedShardList = ColocatedShardIntervalList(shardInterval);
```
[the remark this answers is not in the document]
Comment (SaitTalhaNisanci) [reply src/backend/distributed/operations/shard_rebalancer.c:454]: Yeah, this change shouldn't have been done, nice catch

File: src/backend/distributed/metadata/metadata_utility.c (hunk 18 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/metadata/metadata_utility.c:150]: I think it was just a missing `!`, I don't think it was your local machine :)

File: src/backend/distributed/operations/shard_cleaner.c (hunk 23 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/operations/shard_cleaner.c:78]: I meant with the original PR there were some other changes and `-1` was a leftover from there, now we don't this need anymore

File: src/backend/distributed/operations/repair_shards.c [hunk 32]
```diff
@@ -287,7 +400,78 @@ citus_move_shard_placement(PG_FUNCTION_ARGS)
...
+	int numOfDroppedShards = DropMarkedShards(waitForCleanupLock);
```
[the remark this answers is not in the document]
Comment (SaitTalhaNisanci) [reply src/backend/distributed/operations/repair_shards.c:454]: I am deferring adding the notice until we enable `deferring shards deletion` by default, the other points make sense.

File: src/backend/distributed/metadata/metadata_utility.c [hunk 33]
```diff
@@ -98,12 +101,138 @@ static void AppendShardSizeMinMaxQuery(StringInfo selectQuery, uint64 shardId,
...
+
+	char *availableBytesString = PQgetvalue(result, 0, 0);
+	char *totalBytesString = PQgetvalue(result, 0, 1);
+
+	*availableBytes = SafeStringToUint64(availableBytesString);
+	*totalBytes = SafeStringToUint64(totalBytesString);
+
+	PQclear(result);
+	bool raiseErrors = true;
+	ClearResults(connection, raiseErrors);
```
Comment (SaitTalhaNisanci) [inline src/backend/distributed/metadata/metadata_utility.c:230]: This should be ForgetResult too

File: src/backend/distributed/operations/shard_cleaner.c [hunk 34]
```diff
@@ -55,13 +53,15 @@ master_defer_delete_shards(PG_FUNCTION_ARGS)
...
+ * If dropping any of the shards failed this function returns -1, otherwise it
+ * returns the number of dropped shards.
  */
 int
 TryDropMarkedShards(bool waitForCleanupLock)
 {
-	int droppedShardCount = 0;
 	MemoryContext savedContext = CurrentMemoryContext;
-
+	int droppedShardCount = 0;
```
Comment (SaitTalhaNisanci) [inline src/backend/distributed/operations/shard_cleaner.c:64]: This seems unrelated/unnecessary change

File: src/include/distributed/remote_commands.h [hunk 35]
```diff
@@ -11,6 +11,8 @@
 #ifndef REMOTE_COMMAND_H
 #define REMOTE_COMMAND_H
 
+#include "postgres.h"
```
Comment (SaitTalhaNisanci) [inline src/include/distributed/remote_commands.h:14]: not necessary

File: src/backend/distributed/operations/repair_shards.c [hunk 36]
```diff
@@ -158,6 +169,98 @@ master_copy_shard_placement(PG_FUNCTION_ARGS)
...
+							   connection->hostname, connection->port)));
+	}
+
+	uint64 diskAvailableInBytesAfterShardMove = 0;
+	if (diskAvailableInBytes < colocationSizeInBytes)
+	{
+		/*
+		 * even though the space will be less than "0", we set it to 0 for convenience.
+		 */
+		diskAvailableInBytes = 0;
```
Comment (onderkalaci) [inline src/backend/distributed/operations/repair_shards.c:242]: shouldn't this be `diskAvailableInBytesAfterShardMove`?

File: src/backend/distributed/metadata/metadata_utility.c [hunk 37]
```diff
@@ -98,12 +101,137 @@ static void AppendShardSizeMinMaxQuery(StringInfo selectQuery, uint64 shardId,
...
+	values[1] = UInt64GetDatum(totalBytes);
+
+	HeapTuple diskSpaceTuple = heap_form_tuple(tupleDescriptor, values, isNulls);
+
+	return diskSpaceTuple;
+}
+
+
+/*
+ * citus_local_disk_space_stats returns total disk space and available disk
```
Comment (onderkalaci) [inline src/backend/distributed/metadata/metadata_utility.c:140]: very minor: we typically define UDF function before other static functions

File: src/include/distributed/shard_cleaner.h [hunk 38]
```diff
@@ -14,7 +14,10 @@
 /* GUC to configure deferred shard deletion */
 extern int DeferShardDeleteInterval;
 extern bool DeferShardDeleteOnMove;
+extern double DesiredPercentFreeAfterMove;
+extern bool CheckAvailableSpaceBeforeMove;
 
 extern int TryDropMarkedShards(bool waitForCleanupLock);
+extern int DropMarkedShards(bool waitForCleanupLock);
```
Comment (onderkalaci) [inline src/include/distributed/shard_cleaner.h:21]: can we make this static again?

**Review by onderkalaci — APPROVED** [review]
